File: Day93/Topological-sort_Solution.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution
{
	public:
	//Function to return list containing vertices in Topological order. 
	vector<int> topoSort(int V, vector<int> adj[]) 
	{
	    //Solving this problem with Kahn's Algorithm Techninique
	    int inDegree[V]={0};
	    queue<int> q;
	    //Inserting in-degree of all the vertex of graph and also inserting all the vertex
	    //having 0 in-degree into queue
	    for(int i=0;i<V;i++){
	        for(auto it:adj[i]){
	            inDegree[it]+=1;
	        }
	    }
	    
	    //Inserting all the vertex having in-degree 0 into queue
	    for(int i=0;i<V;i++){
	        if(inDegree[i]==0){
	            q.push(i);
	        }
	    }
	    //Implementing BFS Traversal to process all the adjacencies of corresponding vertex
	    vector<int> result;
	    while(q.empty()==false){
	        int node=q.front();
	        q.pop();
	        
	        //Traverse the adjacency of vertex
	        for(auto it:adj[node]){
	            if(inDegree[it]>0){
	                //Decreasing the vertex in-degree by one
	                inDegree[it]-=1;
	                //by decreassing the vertes degree if it became 0 then need to insert
	                //into queue
	                if(inDegree[it]==0){
	                    q.push(it);
	                }
	            }
	        }
	        //insert the current node for topo sort
	        result.push_back(node);
	    }
	    
	    
	    return result;
	}
};

//{ Driver Code Starts.

/*  Function to check if elements returned by user
*   contains the elements in topological sorted form
*   V: number of vertices
*   *res: array containing elements in topological sorted form
*   adj[]: graph input
*/
int check(int V, vector <int> &res, vector<int> adj[]) {
    
    if(V!=res.size())
    return 0;
    
    vector<int> map(V, -1);
    for (int i = 0; i < V; i++) {
        map[res[i]] = i;
    }
    for (int i = 0; i < V; i++) {
        for (int v : adj[i]) {
            if (map[i] > map[v]) return 0;
        }
    }
    return 1;
}

int main() {
    int T;
    cin >> T;
    while (T--) {
        int N, E;
        cin >> E >> N;
        int u, v;

        vector<int> adj[N];

        for (int i = 0; i < E; i++) {
            cin >> u >> v;
            adj[u].push_back(v);
        }
        
        Solution obj;
        vector <int> res = obj.topoSort(N, adj);

        cout << check(N, res, adj) << endl;
    }
    
    return 0;
}
// } Driver Code Ends

File: Day93/Topological-sort_Solution.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.lang.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader read =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());

        while (t-- > 0) {
            ArrayList<ArrayList<Integer>> list = new ArrayList<>();
            String st[] = read.readLine().trim().split("\\s+");
            int edg = Integer.parseInt(st[0]);
            int nov = Integer.parseInt(st[1]);

            for (int i = 0; i < nov; i++)
                list.add(i, new ArrayList<Integer>());

            int p = 0;
            for (int i = 1; i <= edg; i++) {
                String s[] = read.readLine().trim().split("\\s+");
                int u = Integer.parseInt(s[0]);
                int v = Integer.parseInt(s[1]);
                list.get(u).add(v);
            }

            int[] res = new Solution().topoSort(nov, list);

            if (check(list, nov, res) == true)
                System.out.println("1");
            else
                System.out.println("0");
        }
    }
    static boolean check(ArrayList<ArrayList<Integer>> list, int V, int[] res) {
        
        if(V!=res.length)
        return false;
        
        int[] map = new int[V];
        for (int i = 0; i < V; i++) {
            map[res[i]] = i;
        }
        for (int i = 0; i < V; i++) {
            for (int v : list.get(i)) {
                if (map[i] > map[v]) return false;
            }
        }
        return true;
    }
}

// } Driver Code Ends


/*Complete the function below*/


class Solution
{
    //Function to return list containing vertices in Topological order. 
    static int[] topoSort(int V, ArrayList<ArrayList<Integer>> adj) 
    {
        //Finding Topological sort using Kahn's Algorithm
        int inDegree[]= new int[V];
        
        //finding in-degree of all the corresponding vertex in the graph
        for(int i=0;i<V;i++){
            for(int it:adj.get(i)){
                inDegree[it]++;
            }
        }
        
        //Inserting all the vertex having in-degree 0 into queue for BFS Traversal
        Queue<Integer> q=new LinkedList<Integer>();
        
        for(int i=0;i<V;i++){
            if(inDegree[i]==0){
                q.add(i);
            }
        }
        
        int topoSort[]= new int[V];
        int index=0;
        
        //Implementing BFS traversal
        while(q.isEmpty()==false){
            int node=q.peek();
            topoSort[index++]=node;
            q.poll();
            
            for(int it:adj.get(node)){
                inDegree[it]--;
                if(inDegree[it]==0){
                    q.add(it);
                }
            }
        }
        
        return topoSort;
    }
}
